code/datums/components/revenge_ability.dm

```diff
@@ -0,0 +1,55 @@
+/**
+ * Automatically triggers a linked ability at a target who attacks us.
+ * The ability might not necessarily be on our mob.
+ * Make sure that /datum/element/relay_attackers is also present or you'll never receive the triggering signal.
+ */
+/datum/component/revenge_ability
+	dupe_mode = COMPONENT_DUPE_ALLOWED
+	/// The ability to use when we are attacked
+	var/datum/action/cooldown/ability
+	/// Optional datum for validating targets
+	var/datum/targetting_datum/targetting
+	/// Trigger only if target is at least this far away
+	var/min_range
+	/// Trigger only if target is at least this close
+	var/max_range
+	/// Target the ability at ourself instead of at the offender
+	var/target_self
+
+/datum/component/revenge_ability/Initialize(datum/action/cooldown/ability, datum/targetting_datum/targetting, min_range = 0, max_range = INFINITY, target_self = FALSE)
+	. = ..()
+	if (!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+	src.ability = ability
+	src.targetting = targetting
+	src.min_range = min_range
+	src.max_range = max_range
+	src.target_self = target_self
+
+	RegisterSignal(ability, COMSIG_QDELETING, PROC_REF(ability_destroyed))
+
+/datum/component/revenge_ability/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_ATOM_WAS_ATTACKED, PROC_REF(on_attacked))
+
+/datum/component/revenge_ability/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ATOM_WAS_ATTACKED)
+	if (ability)
+		UnregisterSignal(ability, COMSIG_QDELETING)
+	return ..()
+
+/// If we were attacked, get revenge
+/datum/component/revenge_ability/proc/on_attacked(mob/living/victim, atom/attacker)
+	SIGNAL_HANDLER
+	var/atom/ability_user = ability.owner
+	var/distance = get_dist(ability_user, attacker)
+	if (distance < min_range || distance > max_range)
+		return
+	if (targetting && !targetting.can_attack(victim, attacker))
```

Comment:
does `can_attack()` check faction? i couldn't find it anywhere, because it would be really silly if someone added this component to something like a fireball wizard basic mob and they accidentally get smacked by someone in their faciton and then whammo blammo the player dies due to instant murderization component

if it doesn't and this doesn't seem like the right place to do it, i think the component should handle it

Reply:
The targetting datum of most basic mobs will check faction (and does in this case) unless it's explicitly a mob which doesn't care about factions. I put it on here specifically to stop people whacking their own mount with a lasso in order to spur it to make tentacles 😓